File: DataStructures/bit manipulations/atcoder/A70/A70.py
import sys
from collections import deque

def main():
    input = sys.stdin.read
    data = input().split()
    
    idx = 0
    N = int(data[idx])
    idx += 1
    M = int(data[idx])
    idx += 1

    A = list(map(int, data[idx:idx + N]))
    idx += N

    # 初期状態をビットで表現
    start_state = 0
    for i in range(N):
        if A[i] == 1:
            start_state |= (1 << i)

    # 目標状態（すべてON） = 2^N - 1
    goal_state = (1 << N) - 1

    # 各操作をビットマスクで記録
    ops = []
    for _ in range(M):
        x = int(data[idx]) - 1
        y = int(data[idx + 1]) - 1
        z = int(data[idx + 2]) - 1
        idx += 3
        mask = (1 << x) | (1 << y) | (1 << z)
        ops.append(mask)

    # BFS
    visited = [False] * (1 << N)
    queue = deque()
    queue.append((start_state, 0))
    visited[start_state] = True

    while queue:
        state, steps = queue.popleft()

        if state == goal_state:
            print(steps)
            return

        for op in ops:
            next_state = state ^ op
            if not visited[next_state]:
                visited[next_state] = True
                queue.append((next_state, steps + 1))

    # 到達不可能な場合
    print(-1)

if __name__ == "__main__":
    main()

File: DataStructures/bit manipulations/atcoder/A70/A70.js
const fs = require('fs');

function main(input) {
    const lines = input.trim().split('\n');
    const [N, M] = lines[0].split(' ').map(Number);
    const A = lines[1].split(' ').map(Number);

    // 初期状態をビットで表現
    let startState = 0;
    for (let i = 0; i < N; i++) {
        if (A[i] === 1) startState |= 1 << i;
    }

    // 目標状態（すべてON）
    const goalState = (1 << N) - 1;

    // 各操作をビットマスクに変換
    const ops = [];
    for (let i = 0; i < M; i++) {
        const [x, y, z] = lines[2 + i].split(' ').map((n) => parseInt(n) - 1); // 0-indexed
        let mask = (1 << x) | (1 << y) | (1 << z);
        ops.push(mask);
    }

    // BFS
    const visited = new Array(1 << N).fill(false);
    const queue = [[startState, 0]];
    visited[startState] = true;

    while (queue.length > 0) {
        const [state, steps] = queue.shift();

        if (state === goalState) {
            console.log(steps);
            return;
        }

        for (const op of ops) {
            const nextState = state ^ op;
            if (!visited[nextState]) {
                visited[nextState] = true;
                queue.push([nextState, steps + 1]);
            }
        }
    }

    // 到達不能
    console.log(-1);
}

// ファイル入力処理
main(fs.readFileSync('/dev/stdin', 'utf8'));
